src/k8s/pkg/k8sd/api/kubeconfig.go

```diff
@@ -1,25 +1,60 @@
 package api
 
 import (
+	"encoding/json"
 	"fmt"
 	"net/http"
-	"os"
 
 	apiv1 "github.com/canonical/k8s/api/v1"
+	"github.com/canonical/k8s/pkg/k8sd/api/impl"
+	"github.com/canonical/k8s/pkg/k8sd/setup"
+	"github.com/canonical/k8s/pkg/utils"
 	"github.com/canonical/lxd/lxd/response"
 	"github.com/canonical/microcluster/state"
+	"k8s.io/client-go/tools/clientcmd"
 )
 
 func getKubeconfig(s *state.State, r *http.Request) response.Response {
-	// TODO: Render a new kubeconfig instead of reading the existing one
-	//       when the config can be altered via request parameters.
-	config, err := os.ReadFile("/etc/kubernetes/admin.conf")
+	req := apiv1.GetKubeConfigRequest{}
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		return response.BadRequest(fmt.Errorf("failed to parse request: %w", err))
+	}
+
+	// Fetch pieces needed to render an admin kubeconfig: ca, server, token
+	clusterCfg, err := utils.GetClusterConfig(r.Context(), s)
+	if err != nil {
+		return response.InternalError(fmt.Errorf("failed to retrieve cluster config: %w", err))
+	}
+	ca := clusterCfg.Certificates.CACert
+	server := req.Server
+	if req.Server == "" {
+		server = fmt.Sprintf("%s:%d", s.Address().Hostname(), clusterCfg.APIServer.SecurePort)
+	}
+	token, err := impl.GetOrCreateAuthToken(s.Context, s, "kubernetes-admin", []string{"system:masters"})
+	if err != nil {
+		return response.InternalError(fmt.Errorf("failed to get admin token: %w", err))
+	}
+
+	// Convert rendered kubeconfig bytes into yaml with help from clientcmd
+	kubeconfig, err := setup.ClientKubeconfig(token, server, ca)
+	if err != nil {
+		return response.InternalError(fmt.Errorf("failed to get kubeconfig: %w", err))
+	}
+	clientConfig, err := clientcmd.NewClientConfigFromBytes(kubeconfig)
+	if err != nil {
+		return response.InternalError(fmt.Errorf("failed to create ClientConfig: %w", err))
+	}
+	rawConfig, err := clientConfig.RawConfig()
+	if err != nil {
+		return response.InternalError(fmt.Errorf("failed to create RawConfig: %w", err))
+	}
+	yamlConfig, err := clientcmd.Write(rawConfig)
```

Comment:
oh good! i had a `"Why not stop here", string(kubeconfig))` followup for you but failed to leave it here.  anyway, the back-n-forth is only because you said:

> build a new *clientcmdapi.Config object with this info (the same type that is returned by clientcmd.GetConfigFromFileOrDie()

i figured you must have had a reason for wanting `Config` objects in the picture (e.g. some indirect template validation against clientcmdapi).  Anyway, we're on the same page.  I can stop slinging bytes around and whittle this down into like 2 lines.

Reply:
That would have been something like:

```
config := &clientcmd.Config{}
config.CertificateAuthority = ca
config.AuthInfos = // ..
config.Servers = // ...

// render config object to string
```

It would be greatly preferred to the template we have, but keeping the template for now is also fine.

Reply:
alrighty, we're cooking with gas now.  i dropped the template logic in favor of `clientcmd.Config` generator.